File: shared/src/commonMain/kotlin/com/nasrally/nasrally/ChatViewModel.kt
package com.nasrally.nasrally

import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.postgrest.query.filter.FilterOperator
import io.github.jan.supabase.realtime.PostgresAction
import io.github.jan.supabase.realtime.channel
import io.github.jan.supabase.realtime.decodeRecord
import io.github.jan.supabase.realtime.postgresChangeFlow
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.time.Clock
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
class ChatViewModel(private val groupId: String, private val scope: CoroutineScope) {
    private val _messages = MutableStateFlow<List<Message>>(emptyList())
    val messages: StateFlow<List<Message>> = _messages.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val pageSize = 50
    private var currentPage = 0
    private var canLoadMore = true

    init {
        setupRealtimeSubscription()
    }

    private fun setupRealtimeSubscription() {
        val channel = supabase.channel("chat:$groupId")
        val changeFlow = channel.postgresChangeFlow<PostgresAction.Insert>(
            schema = "public"
        ) {
            table = "messages"
            filter("group_id", FilterOperator.EQ, groupId)
        }

        scope.launch(Dispatchers.Default) {
            changeFlow.collect { action ->
                val newMessage = action.decodeRecord<Message>()
                val current = _messages.value.toMutableList()
                if (current.none { it.id == newMessage.id }) {
                    current.add(0, newMessage)
                    _messages.value = current
                }
            }
        }
        scope.launch { channel.subscribe() }
    }

    suspend fun loadMessages(isRefresh: Boolean = false) {
        if (_isLoading.value || (!canLoadMore && !isRefresh)) return
        _isLoading.value = true
        if (isRefresh) { currentPage = 0; canLoadMore = true }

        val from = currentPage * pageSize
        val to = from + pageSize - 1

        try {
            val fetched: List<Message> = supabase.from("messages")
                .select {
                    filter { eq("group_id", groupId) }
                    order("created_at", order = Order.DESCENDING)
                    range(from.toLong(), to.toLong())
                }.decodeList()

            if (fetched.size < pageSize) canLoadMore = false

            val current = if (isRefresh) mutableListOf() else _messages.value.toMutableList()
            current.addAll(fetched)
            _messages.value = current
            currentPage++
        } catch (e: Exception) {
            println("Fetch error: ${e.message}")
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun sendMessage(content: String, senderId: String) {
        val newMessage = Message(
            id = Uuid.random().toString(),
            groupId = groupId,
            senderId = senderId,
            content = content,
            createdAt = Clock.System.now().toString()
        )

        val current = _messages.value.toMutableList()
        current.add(0, newMessage)
        _messages.value = current

        try {
            supabase.from("messages").insert(newMessage)
        } catch (e: Exception) {
            _messages.value = _messages.value.filter { it.id != newMessage.id }
        }
    }
}

File: iosApp/iosApp/ChatViewModel.swift
import SwiftUI
import Supabase
import Combine

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [Message] = []
    @Published var isLoading = false
    
    private let client: SupabaseClient
    let groupId: UUID
    private let pageSize = 50
    private var currentPage = 0
    private var canLoadMore = true
    private var channel: RealtimeChannelV2?
    
    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()
    
    init(client: SupabaseClient, groupId: UUID) {
        self.client = client
        self.groupId = groupId
        setupRealtimeSubscription()
        Task {
            for await status in client.realtimeV2.statusChange {
                print("Realtime Status: \(status)")
            }
        }
    }
    
    private func setupRealtimeSubscription() {
        let channel = client.channel("chat:\(groupId.uuidString)")
        
        // Listen for new messages
        let subscription = channel.postgresChange(
            InsertAction.self,
            schema: "public",
            table: "messages",
            filter: .eq("group_id", value: groupId.uuidString)
        )
        
        Task {
            for await action in subscription {
                do {
                    let newMessage = try action.decodeRecord(as: Message.self, decoder: self.decoder)
                    await MainActor.run {
                        if !self.messages.contains(where: { $0.id == newMessage.id }) {
                            self.messages.insert(newMessage, at: 0)
                        }
                    }
                } catch { print("Realtime decode error: \(error)") }
            }
        }
        
        Task { try? await channel.subscribeWithError() }
        self.channel = channel
    }

    func loadMessages(isRefresh: Bool = false) async {
        guard !isLoading && (canLoadMore || isRefresh) else { return }
        isLoading = true
        if isRefresh { currentPage = 0; canLoadMore = true }
        
        let from = currentPage * pageSize
        let to = from + pageSize - 1
        
        do {
            let fetchedMessages: [Message] = try await client.from("messages")
                .select()
                .eq("group_id", value: groupId)
                .order("created_at", ascending: false)
                .range(from: from, to: to)
                .execute()
                .value
            
            if fetchedMessages.count < pageSize { canLoadMore = false }
            
            await MainActor.run {
                if isRefresh { self.messages = fetchedMessages } 
                else { self.messages.append(contentsOf: fetchedMessages) }
                currentPage += 1
            }
        } catch { print("Fetch error: \(error)") }
        isLoading = false
    }
    
    func sendMessage(_ content: String, senderId: UUID) async {
        let newMessage = Message(
            id: UUID(),
            groupId: groupId,
            senderId: senderId,
            content: content,
            createdAt: Date()
        )
        
        // Add locally immediately so UI doesn't look broken
        await MainActor.run {
            self.messages.insert(newMessage, at: 0)
        }
        
        do {
            try await client.from("messages")
                .insert(newMessage)
                .execute()
        } catch {
            print("Insert Error: \(error)")
            // Remove locally if it fails
            await MainActor.run {
                self.messages.removeAll(where: { $0.id == newMessage.id })
            }
        }
    }
}
